README.md
# @janiscommerce/totem-checkout


## Installation
```sh
npm install @janiscommerce/totem-checkout
```

## API


## Usage
```js
const { Payment, Cart, Order } = require('@janiscommerce/totem-checkout');

const cartInstance = new Cart({ environment: 'tiendaqa' })

const orderInstance = new Order({ environment: 'tiendaqa', apiKey: 'API_KEY', apiToken: 'API_TOKEN' });

const paymentInstance = new Payment({ environment: 'tiendaqa', apiKey: 'API_KEY', apiToken: 'API_TOKEN', slaId: 'SLA_ID' });
```

## Cart

## Methods

## addItems(items) ⇒ `Promise`
Function for add a new item or items to current order

| Param          | Type             | Description                                      |
| ---            | ---              | ---                                              |
| items          | `array`/`object` | Object or Array with Object with id and quantity |
| items.id       | `number`         | id of sku product.                               |
| items.quantity | `number`         | quantity items                                   |

**Example**
```js
const cart = await cartInstance.addItems({
	id: 45,
	quantity: 2
});

const cart = await cartInstance.addItems([{
	id: 45,
	quantity: 2
},{
	id: 45,
	quantity: 2
}]);
```

## changeItemQuantity(items) ⇒ `Promise`
Function for change the quantity to item into the current orderForm

| Param          | Type             | Description                                      |
| ---            | ---              | ---                                              |
| items          | `array`/`object` | Object or Array with Object with id and quantity |
| items.id       | `number`         | id of sku product.                               |
| items.quantity | `number`         | quantity items                                   |

**Example**
```js
const cart = await cartInstance.changeItemsQuantity({
	id: 45,
	quantity: 2
});

const cart = await cartInstance.changeItemsQuantity([{
	id: 45,
	quantity: 5
},{
	id: 45,
	quantity: 8
}]);
```

## removeItems(items) ⇒ `Promise`
Function for remove items

| Param          | Type             | Description                                      |
| ---            | ---              | ---                                              |
| items          | `array`/`object` | Object or Array with Object with id and quantity |
| items.id       | `number`         | id of sku product.                               |

**Example**
```js
const cart = await cartInstance.removeItems({
		id: 44
});

const cart = await cartInstance.removeItems([{
	id: 45
},{
	id: 46
}]);
```


## removeAllItems() ⇒ `Promise`
Function for remove all items
**Example**
```js
const cart = await CartInstance.removeAllItems();
```

## simulate(items) ⇒ `Promise`
Function for simulate cart

| Param          | Type             | Description                                      |
| ---            | ---              | ---                                              |
| items          | `array`/`object` | Object or Array with Object with id and quantity |
| items.id       | `number`         | id of sku product.                               |
| items.quantity | `number`         | quantity items                                   |
| postalCode     | `number`         | PostalCoode for get availables slas              |

**Example**
```js
const cartSimulated = await cartInstance.simulate({
    id: 17,
    quantity: 2
}, 1000);

 const cartSimulated = await cartInstance.simulate([{
    id: 17,
    quantity: 2
}], 1000);
```

## clearCart()
Function for clear cart cached in instance then call getOrder method
**Example**
```js
await cartInstance.getCart();
// into cartInstance.cart getCart method save cart created

cartInstance.clearCart();
// cartInstance.cart is null
```


## Order

## Methods

## create(data) ⇒ `Promise`
Create new Order

| Param                | Type              | Description                                      |
| ---                  | ---               | ---                                              |
| data                 | `object`          | Object with data necesary for make order         |
| data.cart            | `object`          | Object Cart or Object Cart simulation            |
| data.slaId           | `string`/         | logistic sla id                                  |
| data.paymentSystemId | `number`/`string` | id of payment system                             |

**If order is created, the order response and passed data is cached in instance**

**Example**
```js
// Create order with payment promisory
await orderInstance.create({
	cart: orderSimulated,
	slaId: 'Normal',
	paymentSystemId: 201
});

// Create order with payment card
await orderInstance.create({
	cart: orderSimulated,
	slaId: 'Normal',
	paymentSystemId: 201,
	paymentFields: {
		holderName: 'Test TEST',
		cardNumber: '4700000000000000',
		validationCode: '',
		dueDate: "12/22",
		document: '',
		accountId: '',
		address: null,
		callbackUrl: ''
	}
});
```

## sendTransactionPayments(data, transactionId) ⇒ `Promise`
Send to order created the paymentData

| Param                | Type              | Description                                      |
| ---                  | ---               | ---                                              |
| data                 | `object`          | Object with data necesary for make order         |
| data.cart            | `object`          | Object Cart or Object Cart simulation            |
| data.slaId           | `string`/         | logistic sla id                                  |
| data.paymentSystemId | `number`/`string` | id of payment system                             |
| transactionId        | `string`          | id of transaction                                |

**if not pass data and transactionId, the function use data passed in method create, saved in instance**

**Example**
```js
// Send transaction custom
const data = {
	cart: orderSimulated,
	slaId: 'Normal',
	paymentSystemId: 201
};

await orderInstance.sendTransactionPayments(data, 'TRANSACTION_ID');

// send transaction with data and transaction id cached
await orderInstance.sendTransactionPayments();
```

## authorizeTransaction(data) ⇒ `Promise`
Authorize Payment Methods order

| Param                | Type              | Description                                      |
| ---                  | ---               | ---                                              |
| transactionId        | `string`          | id of transaction                                |

**if not pass transactionId, the function use transactionId in cached order, saved in instance**

**Example**
```js
// authorizeTransaction custom
await orderInstance.authorizeTransaction('TRANSACTION_ID');

// authorizeTransaction with transactionId in order cached
await orderInstance.authorizeTransaction();
```

## clearOrder()
Clear order created cached and data initial passed
**Example**
```js
await orderInstance.create({
	cart: orderSimulated,
	slaId: 'Normal',
	paymentSystemId: 201
});
// into orderInstance.order, create method save order created
// Into orderInstace.initialdata, create method save argument data passed

orderInstance.clearOrder();
// orderInstance.order is null
// orderInstance.initialData is null
```

## Creating a Order

**Example for simulate cart, and then create cart and send and authorize payment**
```js
const { Cart, Order } = require('totem-app-checkout');

const evironment = 'testqa';
const cart = new Cart({ environment });

const order = new Order({
	environment,
	apiKey: 'API_KEY',
	apiToken: 'API_TOKEN'
});

const items = [
	{ id: 8, quantity: 1 },
	{ id: 7, quantity: 1 }
];

const cartSimulated = await cart.simulate(items, 1000)


const data = await order.create({
	cart: orderSimulated,
	slaId: 'Normal',
	paymentSystemId: 201,
	paymentFields: {
		holderName: 'Test VTEX',
		cardNumber: '4704550000000005',
		validationCode: '',
		dueDate: "12/22",
		document: '',
		accountId: '',
		address: null,
		callbackUrl: ''
	}
});

const sendPaymentResponse = await order.sendTransactionPayments();

const authorizeResponse = await order.authorizeTransaction();
```

## Payment

## Methods

## collect(value, paymentSystem, transactionId, cart, isCash, isContactless) ⇒ `Promise`
Function for add a new item or items to current order

| Param          | Type             | Description                                      | Default |
| ---            | ---              | ---                                              | ---     |
| value          | `number`         | Value cart total                                 |         |
| paymentSystem  | `string|number`  | id of payment                                    |         |
| transactionId  | `string`         | id of transaction                                |         |
| cart           | `object`         | object cart simulated                            |         |
| isCash         | `boolean`        | is payment in cash or not                        | false   |
| isContactless  | `boolean`        | is payment with card contacless                  | false   |

## Creating a payment

```js
  // payment with transaction id with card
  await paymentInstance.collect(18300, 2, 'TRANSACTION_ID');

  // payment with cart simlated with card
  await paymentInstance.collect(18300, 2, null, cartSimulatedObj);

  // payment with transaction id in cash
  await paymentInstance.collect(18300, 1, 'TRANSACTION_ID', null, true);

  // payment with cart simlated in cash
  await paymentInstance.collect(18300, 1, null, cartSimulatedObj, true);
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.1.0] - 2019-10-28
- Add fullpayment with mt188-card-reader and vtex
- Fix card reader execution
- isContactless option added

## [1.0.0] - 2019-09-20
### Added
- First package version


lib/cart.js
'use strict';

class Cart {

	constructor({
		environment,
		seller = '1',
		sc = 1
	} = {}) {
		this.environment = environment;
		this.urlBase = `https://${environment}.myvtex.com`;
		this.sc = sc;
		this.seller = seller;
		this.urlPath = '/api/checkout/pub/orderForm';
		this.cart = null;

		this.valitateParams();
	}


	static get itemsTypeError() {
		return 'Only accepts object or array with objects';
	}


	static get expectedOrderFormSections() {
		return {
			expectedOrderFormSections: [
				'items',
				'totalizers',
				'clientProfileData',
				'shippingData',
				'paymentData',
				'sellers',
				'messages',
				'marketingData',
				'clientPreferencesData',
				'storePreferencesData',
				'giftRegistryData',
				'ratesAndBenefitsData',
				'openTextField',
				'commercialConditionData',
				'customData'
			]
		};
	}

	// Get orderId fro current instance order
	get cartId() {
		return this.cart && this.cart.orderFormId;
	}


	valitateParams() {
		if(!this.environment)
			throw new Error('environment not defined, please pass environment in instance');
	}


	makeUrl(urlModifier, cartId) {
		const id = cartId || this.cartId;
		const base = `${this.urlBase}${this.urlPath}`;
		return `${base}${id ? `/${id}` : ''}${urlModifier}?sc=${this.sc}`;
	}


	async call(url, options) {
		const headers = { 'Content-Type': 'application/json' };

		const res = await fetch(url, { ...options, headers });

		if(!res.ok)
			throw await res.json();

		return res.json();
	}


	makeItemsArray(items) {
		const orderItems = [];

		if(!(items instanceof Object && typeof items !== 'function'))
			throw new Error(Cart.itemsTypeError);

		if(Array.isArray(items)) {

			if(!items.every(item => item instanceof Object && typeof item !== 'function' && !Array.isArray(item)))
				throw new Error(Cart.itemsTypeError);

			return items.map(item => {
				return {
					...item,
					seller: this.seller
				};
			});
		}

		orderItems.push({ ...items, seller: this.seller });

		return orderItems;
	}


	/**
	 * @method getCart
	 * @description Get an existing cart or create a new one
	 * @param {string} cartId
	 * @private
	 * @example
	 *
	 * const cart = await CartInstance.getCart(); // create cart
	 * const cart = await CartInstance.getCart('asd654ad654asd564asd'); // get an existing cart
	 *
	 */
	async getCart(cartId) {
		const url = this.makeUrl('', cartId);

		try {
			const response = await this.call(url, { method: 'POST' });

			this.cart = response;

			return response;
		} catch(error) {
			throw error;
		}
	}


	/**
	 * @method addItems
	 * @description Function for add a new item or items to current order
	 * @param {array|object} items Object or Array of Objects
	 * @param {number} items.id id of sku product.
 	 * @param {number} items.quantity
	 * @returns {Promise}
	 * @example
	 *
	 * 	const cart = await CartInstance.addItems({
	 * 		id: 45,
	 * 		quantity: 2
	 * });
	 *
	 * const cart = await CartInstance.addItems([{
	 * 		id: 45,
	 * 		quantity: 2
	 * },{
	 * 		id: 45,
	 * 		quantity: 2
	 * }]);
	 *
	*/
	async addItems(items) {
		const url = this.makeUrl('/items');

		try {

			const orderItems = this.makeItemsArray(items);

			const body = JSON.stringify({
				orderItems,
				...Cart.expectedOrderFormSections
			});

			const response = await this.call(url, { method: 'POST', body });

			this.cart = response;

			return response;
		} catch(error) {
			throw error;
		}
	}


	/**
	 * @method updateItems
	 * @description Helper for update item in multiple forms
 	 * @param {array|object} items Object or Array of Objects
	 * @returns {Promise}
	 */
	updateItems(items) {
		if(!this.cart)
			throw new Error('Not exist cart data, please call getCart() for create one');

		const orderItems = items.map(item => {
			const index = this.cart.items.findIndex(itm => parseInt(itm.id, 10) === parseInt(item.id, 10));

			return {
				index,
				hasBundleItems: false,
				...item
			};
		});

		const url = this.makeUrl('/items/update');

		const body = JSON.stringify({
			orderItems,
			noSplitItem: true,
			...Cart.expectedOrderFormSections
		});

		return this.call(url, { method: 'POST', body });
	}


	/**
	 * @method changeItemQuantity
	 * @description Function for change the quantity to item into the current orderForm
 	 * @param {array|object} items Object or Array of Objects
	 * @param {number} items.id id of sku product.
 	 * @param {number} items.quantity
	 * @returns {Promise}
	 * @example
	 *
	 * 	const cart = await CartInstance.changeItemsQuantity({
	 * 		id: 45,
	 * 		quantity: 2
	 * });
	 *
	 * const cart = await CartInstance.changeItemsQuantity([{
	 * 		id: 45,
	 * 		quantity: 5
	 * },{
	 * 		id: 45,
	 * 		quantity: 8
	 * }]);
	 *
	 */
	async changeItemsQuantity(items) {
		try {
			const orderItems = this.makeItemsArray(items);

			const response = await this.updateItems(orderItems);

			this.cart = response;

			return response;
		} catch(error) {
			throw error;
		}
	}


	/**
	 * @method removeItems
	 * @description Function for remove items
 	 * @param {array|object} items Object or Array of Objects
	 * @param {number} items.id id of sku product
	 * @returns {Promise}
	 * @example
	 *
	 * 	const cart = await CartInstance.removeItems({
	 * 		id: 44
	 * });
	 *
	 * const cart = await CartInstance.removeItems([{
	 * 		id: 45
	 * },{
	 * 		id: 46
	 * }]);
	 *
	 */
	async removeItems(items) {
		try {
			const itemsMaked = this.makeItemsArray(items);

			const orderItems = itemsMaked.map(item => {
				return {
					...item,
					quantity: 0
				};
			});

			const response = await this.updateItems(orderItems);

			this.cart = response;

			return response;
		} catch(error) {
			throw error;
		}
	}


	/**
	 * @method removeAllItems
	 * @description Function for remove all items
	 * @returns {Promise}
	 * @example
	 *
	 * 	const cart = await CartInstance.removeAllItems();
	 *
	 */
	async removeAllItems() {
		if(!this.cart)
			throw new Error('Not exist cart data, please call getCart() for create one');

		try {
			const items = this.cart.items.map(item => ({ id: item.id }));

			const response = await this.removeItems(items);

			this.cart = response;

			return response;
		} catch(error) {
			throw error;
		}
	}


	/**
	 * @method simulate
	 * @description Function for simulate cart
 	 * @param {array|object} items Object or Array of Objects
	 * @param {number} items.id id of sku product.
 	 * @param {number} items.quantity
	 * @returns {Promise}
	 * @example
	 *
	 * const cartSimulated = await CartInstance.simulate({
	 *     id: 17,
	 *     quantity: 2
	 * });
	 *
	 *  const cartSimulated = await CartInstance.simulate([{
	 *     id: 17,
	 *     quantity: 2
	 * }]);
	 *
	 */
	simulate(items, postalCode) {
		const orderItems = this.makeItemsArray(items);

		const data = {
			items: orderItems,
			postalCode,
			country: 'ARG'
		};

		const url = `${this.urlBase}${this.urlPath.toLowerCase()}s/simulation`;

		return this.call(url, { method: 'POST', body: JSON.stringify(data) });
	}


	/**
	 * @method clearCart
	 * @description Clear cart cached
	 */
	clearCart() {
		this.cart = null;
	}
}

module.exports = Cart;
